syntax = "proto3";
package crypto;
service CryptoService {
    rpc bootSystem(BootSystemRequest) returns (Empty);
    rpc addProccess(AddProcessRequest) returns (Empty);
    rpc configure(ConfigureRequest) returns (Empty);
    rpc generateAESKey(GenerateAESKeyRequest) returns (GenerateAESKeyResponse);
    rpc generateRSAKeyPair(GenerateKeyPairRequest) returns (GenerateKeyPairResponse);
    rpc generateECCKeyPair(GenerateKeyPairRequest) returns (GenerateKeyPairResponse);
    rpc getSignedDataLength(GetHashLengthRequest) returns (GetLengthResponse);
    rpc getECCencryptedLength(GetLengthRequest) returns (GetLengthResponse);
    rpc getECCDecryptedLength(GetLengthRequest) returns (GetLengthResponse);
    rpc getRSAencryptedLength(GetLengthRequest) returns (GetLengthResponse);
    rpc getRSAdecryptedLength(GetLengthRequest) returns (GetLengthResponse);
    rpc getAESencryptedLength(GetAESLengthRequest) returns (GetLengthResponse);
    rpc getAESdecryptedLength(GetAESLengthRequest) returns (GetLengthResponse);
    rpc getEncryptedLen(GetWholeLength) returns (GetLengthResponse);
    rpc getDecryptedLen(GetWholeLength) returns (GetLengthResponse);
    rpc sign(SignRequest) returns (SignResponse);
    rpc verify(VerifyRequest) returns (VerifyResponse);
    rpc getPublicECCKeyByUserId(KeyRequest) returns (KeyResponse);
    rpc getPublicRSAKeyByUserId(KeyRequest) returns (KeyResponse);
    rpc ECCencrypt(AsymetricEncryptRequest) returns (AsymetricEncryptResponse);
    rpc ECCdecrypt(AsymetricDecryptRequest) returns (AsymetricDecryptResponse);
    rpc RSAencrypt(AsymetricEncryptRequest) returns (AsymetricEncryptResponse);
    rpc RSAdecrypt(AsymetricDecryptRequest) returns (AsymetricDecryptResponse);
    rpc AESencrypt(AESEncryptRequest) returns (AESEncryptResponse);
    rpc AESdecrypt(AESDecryptRequest) returns (AESDecryptResponse);
    rpc encrypt(EncryptRequest) returns (EncryptResponse);
    rpc decrypt(DecryptRequest) returns (DecryptResponse);
    rpc signUpdate(SignRequest) returns (SignResponse);
    rpc signFinalize(SignRequest) returns (SignResponse);
    rpc verifyUpdate(VerifyRequest) returns (VerifyResponse);
    rpc verifyFinalize(VerifyRequest) returns (VerifyResponse);
}
message AsymetricEncryptRequest{
    int32 senderId = 1;
    string keyId = 2;
    bytes data = 3;
}
message AsymetricEncryptResponse{
    bytes encrypted_data = 1;
}
message AsymetricDecryptRequest{
    int32 receiverId = 1;
    string keyId = 2;
    bytes data = 3;
}
message GetHashLengthRequest{
    SHAAlgorithm func = 1;
    int32 dataLen = 2;
}
message GetAESLengthRequest{
   int32 dataLen = 1;
   bool isFirst = 2;
   AESChainingMode chainingMode = 3;
}
message AsymetricDecryptResponse{
    bytes decrypted_data = 1;
}
message GetLengthRequest {
   int32 in_len = 1;
}
message GetLengthResponse {
   int32 len = 1;
}
message GetWholeLength{
    int32 senderId = 1;
     int32 inLen = 2;
}
message GenerateAESKeyRequest {
    int32 user_id = 1;
    repeated KeyPermission permissions = 2;
    AESKeyLength keyLength = 3;
    int32 destUserId = 4;
}
message GenerateAESKeyResponse {
    string aes_key = 1;
}
message GenerateKeyPairRequest {
    int32 user_id = 1;
    repeated KeyPermission permissions = 2;
}
message GenerateKeyPairResponse {
    string public_key = 1;
    string private_key = 2;
}
message SignRequest {
    int32 sender_id = 1;
    bytes data = 2;
    SHAAlgorithm hash_func = 3;
    int64 counter = 5;
    string key_id = 6;
}
message SignResponse {
    bytes signature = 1;
}
message VerifyRequest {
    int32 sender_id = 1;
    int32 receiver_id = 2;
    bytes data = 3;
    bytes signature = 4;
    SHAAlgorithm hash_func = 5;
    string key_id = 6;
    int32 counter = 7;
}
message VerifyResponse {
    bool valid = 1;
    bytes out = 2;
}
message KeyRequest {
    int32 user_id = 1;
}
message KeyResponse {
    string key = 1;
}
message UserKeyPermissions {
  int32 userId = 1;
  repeated KeyPermission permissions = 2;
}
message BootSystemRequest {
  repeated UserKeyPermissions usersIdsPermissions = 1;
}
message Empty{
}
message CryptoConfig {
  SHAAlgorithm hashFunction = 1;
  AESKeyLength aesKeyLength = 2;
  AESChainingMode aesChainingMode = 3;
  AsymmetricFunction asymmetricFunction = 4;
}
message ConfigureRequest {
  int32 userId = 1;
  CryptoConfig config = 2;
}
message AddProcessRequest{
    int32 userId = 1;
    repeated KeyPermission permissions = 2;
}
message EncryptRequest {
    int32 sender_id = 1;
    int32 receiver_id = 2;
    bytes data = 3;
    int64 counter = 4;
    bool isFirst = 5;
}
message EncryptResponse {
    bytes encrypted_data = 1;
    bytes signature = 2;
}
message DecryptRequest {
    int32 sender_id = 1;
    int32 receiver_id = 2;
    bytes encrypted_data = 3;
    int64 counter = 4;
    bytes signature = 5;
    bool isFirst = 6;
}
message DecryptResponse {
    bytes decrypted_data = 1;
}
message AESEncryptRequest {
    int32 sender_id = 1;
    int32 receiver_id = 2;
    bytes data = 3;
    AsymmetricFunction func = 4;
    int64 counter = 5;
    string key_id = 6;
    AESKeyLength key_length = 7;
    AESChainingMode chainingMode = 8;
}
message AESEncryptResponse {
    bytes encrypted_data = 1;
}
message AESDecryptRequest {
    int32 sender_id=1;
    int32 receiver_id = 2;
    bytes data_in = 3;
    int32 in_len=4;
    bytes data_out = 5;
    AsymmetricFunction func = 6;
    AESKeyLength key_length = 7;
    AESChainingMode chainingMode=8;
    int64 counter = 9;
    string key_id = 10;
}
message AESDecryptResponse {
    bytes decrypted_data = 1;
}
enum KeyPermission {
    VERIFY = 0;
    SIGN = 1;
    ENCRYPT = 2;
    DECRYPT = 3;
    EXPORTABLE = 4;
}
enum AESChainingMode {
  ECB = 0; 
  CBC = 1; 
  CFB = 2; 
  OFB = 3;
  CTR = 4;
}
enum AsymmetricFunction {
    RSA = 0;
    ECC = 1;
}
enum SHAAlgorithm {
    SHA256 = 0;
    SHA3_512 = 1;
}
enum AESKeyLength {
    AES_128 = 0;
    AES_192 = 1;
    AES_256 = 2;
}
